cli/src/commands/rebase.rs

```diff
@@ -873,17 +874,33 @@ fn move_commits(
             if let Some(new_child_parents) = new_children_parents.get(commit_id) {
                 new_child_parents.clone()
             }
-            // Commits in the target set should persist only rebased parents from the target
-            // sets.
-            else if let Some(target_commit_parents) =
-                target_commits_internal_parents.get(commit_id)
-            {
-                // If the commit does not have any parents in the target set, it is one of the
-                // commits in the root set, and should be rebased onto the new destination.
-                if target_commit_parents.is_empty() {
+            // Commit is in the target set.
+            else if target_commit_ids.contains(commit_id) {
+                // If the commit is a root of the target set, it should be rebased onto the new destination.
+                if target_roots.contains(commit_id) {
                     new_parent_ids.clone()
-                } else {
-                    target_commit_parents.clone()
+                }
+                // Otherwise:
+                // 1. Keep parents which are within the target set.
+                // 2. Replace parents which are outside the target set but are part of the
+                //    connected target set with their ancestor commits which are in the target
+                //    set.
+                // 3. Keep other parents outside the target set if they are not descendants of the
+                //    new children of the target set.
+                else {
+                    let mut new_parents = vec![];
+                    for parent_id in commit.parent_ids() {
+                        if target_commit_ids.contains(parent_id) {
+                            new_parents.push(parent_id.clone());
+                        } else if let Some(parents) =
+                                connected_target_commits_internal_parents.get(parent_id) {
+                            new_parents.extend(parents.iter().cloned());
+                        } else if !new_children.iter().any(|new_child| {
+                                mut_repo.index().is_ancestor(new_child.id(), parent_id) }) {
+                            new_parents.push(parent_id.clone());
+                        }
+                    }
+                    new_parents
```

Comment:
It seems a bit odd that the target root is special cased (even if it was a merge.)

For example, rebasing `d::e` off `a|b` removes `a|b` from `d`, but not from `e`. Isn't it better to exclude all ancestors of the target roots?

```
    create_commit(&test_env, &repo_path, "a", &[]);
    create_commit(&test_env, &repo_path, "b", &[]);
    create_commit(&test_env, &repo_path, "c", &[]);
    create_commit(&test_env, &repo_path, "d", &["a", "b"]);
    create_commit(&test_env, &repo_path, "e", &["d", "a", "b"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @      e
    ├─┬─╮
    ◉ │ │  d
    ╰─┬─╮
      │ ◉  b
      ◉ │  a
      ├─╯
    ◉ │  c
    ├─╯
    ◉
    "###);

    test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "d::e", "-d", "c"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @      e
    ├─┬─╮
    │ │ ◉  b
    │ ◉ │  a
    │ ├─╯
    ◉ │  d
    ◉ │  c
    ├─╯
    ◉
    "###);
```

Reply:
Yeah, I can see where you're coming from, it might also makes sense to treat all the revisions passed as an independent set and move it directly to the destination whilst removing any links to other commits. Here's the thread in Discord where I discussed this: https://discord.com/channels/968932220549103686/1234553998703460423.

The example described shows that with the current behavior, rebasing `-r s::` would differ from `-s s`:

```
❯ jj l
@    vwysomoy benjamin@dev.ofcr.se 1 second ago 5331ba13
├─╮  (empty) d
○ │  kurkunwm benjamin@dev.ofcr.se 30 seconds ago 8d3b09e0
│ │  (empty) c
○ │  svotuppn benjamin@dev.ofcr.se 34 seconds ago 001cebf5
├─╯  (empty) b
○  ksnsnuxr benjamin@dev.ofcr.se 43 seconds ago e55b9272
│  (empty) a
◆  zzzzzzzz root() 00000000

❯ jj rebase -r s:: -d ks
Skipping rebase of commits:
  svotuppn 001cebf5 (empty) b
  kurkunwm 8d3b09e0 (empty) c
Rebased 1 commits onto destination
Working copy now at: vwysomoy ac20feeb (empty) d
Parent commit      : kurkunwm 8d3b09e0 (empty) c

❯ jj l
@  vwysomoy benjamin@dev.ofcr.se 1 second ago ac20feeb
│  (empty) d
○  kurkunwm benjamin@dev.ofcr.se 1 minute ago 8d3b09e0
│  (empty) c
○  svotuppn benjamin@dev.ofcr.se 1 minute ago 001cebf5
│  (empty) b
○  ksnsnuxr benjamin@dev.ofcr.se 1 minute ago e55b9272
│  (empty) a
◆  zzzzzzzz root() 00000000

❯ jj undo
Working copy now at: vwysomoy 5331ba13 (empty) d
Parent commit      : kurkunwm 8d3b09e0 (empty) c
Parent commit      : ksnsnuxr e55b9272 (empty) a

❯ jj rebase -s s -d ks
Skipping rebase of commit svotuppn 001cebf5 (empty) b

❯ jj l
@    vwysomoy benjamin@dev.ofcr.se 1 minute ago 5331ba13
├─╮  (empty) d
○ │  kurkunwm benjamin@dev.ofcr.se 1 minute ago 8d3b09e0
│ │  (empty) c
○ │  svotuppn benjamin@dev.ofcr.se 1 minute ago 001cebf5
├─╯  (empty) b
○  ksnsnuxr benjamin@dev.ofcr.se 1 minute ago e55b9272
│  (empty) a
◆  zzzzzzzz root() 00000000
```

I thought that might be a little inconsistent, but as @martinvonz pointed out we can also explicitly make the behavior different for `-r` and `-s`/`-b`.

Any other opinions on the desired behavior for this?

Reply:
To @yuja 's comment, I think of it in terms of a more realistic example:

```
@    f
├─╮
│ ○  e
│ ○  d
○ │  c
○ │  b
├─╯
○  a
◆
```

There's a main branch `b` and `c` and a feature branch `d` and `e` that got merged as `f`. If we were to rebase `c` and `f` to some unrelated part of the commit graph, I would expect `f` to still have `a` as an ancestor (through `e` and `d`).

I don't think the logic should change if `a` were instead a direct parent of `f` (that is, `e` and `d` are missing).

Regarding @bnjmnt4n 's question, I did not read the Discord thread, but think that `-r s::` should have the same behavior to `-s s` if possible.

Reply:
> The example described shows that with the current behavior, rebasing `-r s::` would differ from `-s s`:

I see. Even though it's inconsistent, I agree that making `-r s::` identical to `-s s` is practically useful.
Perhaps, that's because `-r` specifies nodes to rebase, whereas `-s` specifies edges to break. We need to deduce edges from `-r` sets, so the ambiguity is unavoidable.

Can you include the reason behind this change in the commit message?